dap-mode.el

```diff
@@ -1493,6 +1493,39 @@ has language id = LANGUAGE-ID. The function must return debug
 arguments which contain the debug port to use for opening TCP connection."
   (puthash language-id provide-configuration-fn dap--debug-providers))
 
+(cl-defun dap-register-setup-debug-provider (&rest debug-setup
+						 &key
+						 language-id
+						 process-template-fn
+						 present?
+						 install)
+  "Register debug setup for LANGUAGE-ID.
+
+The only purpose of DEBUG-SETUP is to bind to `&key' passed as arguments.
+LANGUAGE-ID is the language ID to identify the LANGUAGE.
+PROCESS-TEMPLATE-FN is a function that procecess the template used by the
+debugger.
+PRESENT? is a function that checks if the debug server is installed or 
+outdated.
+INSTALL is a function that performs the installation of the DAP SERVER."
+
+  ;; I used 'cond' because All the tested args are members of
+  ;; DEBUG-SETUP.  I tested them directly instead of doing the check
+  ;; with ((null (plist-member ... debug-setup)) (error ...)) because
+  ;; it makes more sens.
+  (cond
+   ((null debug-setup)         (error (concat "'dap-register-setup-debug-provider' takes keyword arguments.\n"
+					      "Please refer to it documentation")))
+   ((null language-id)         (error "'langauge-id' is not provided"))
+   ((null process-template-fn) (error "'process-template-fn' is not provided"))
+   ((null present?)            (error "'present?' is not provided"))
+   ((null install)             (error "'install' is not provided"))
+   (t nil))
+  
+  (let ((alist-debug-setup (seq-partition debug-setup 2)))
+    (dolist (entry alist-debug-setup)
+      (puthash (car entry) (cadr entry) dap--debug-providers))))
```

Comment:
I think we should either use language-id as a key and associate it to plist/alist with the other keys or use composite key something like `(cons language-id :present?)` because when we register 2 providers we will have 2 entries in that alist and there won't be a way to distinguish what is the corresponding :present? function, right?

Reply:
I would go for association of #( ... :language-id plist) it make more sense, we just ask for the key in the plist and we don't expect duplication of keys and values. for the composite key, i don't understand your suggestion. it would be something like this ((language-id :present) . present?) ?.